lib/wibox/layout/overflow.lua

```diff
@@ -0,0 +1,408 @@
+    if spacing_widget then
+        if is_y then
+            local _
+            _, spacing = base.fit_widget(self, context, spacing_widget, width, spacing)
+        else
+            spacing = base.fit_widget(self, context, spacing_widget, spacing, height)
+        end
+    end
+
+    for i, w in pairs(widgets) do
+        local content_x, content_y
+        local content_w, content_h = base.fit_widget(self, context, w, width, height)
+
+        -- When scrolling down, the content itself moves up -> substract
+        local scrolled_pos = pos - (scroll_position * interval)
+
+        -- Stop processing completely once we're passed the visible portion
+        if scrolled_pos > avail_in_dir then
+            break
+        end
+
+        if is_y then
+            content_x, content_y = widget_x, scrolled_pos
+            pos = pos + content_h + spacing
+
+            if self._private.fill_space then
+                content_w = width
+            end
+        else
+            content_x, content_y = scrolled_pos, widget_y
+            pos = pos + content_w + spacing
+
+            if self._private.fill_space then
+                content_h = height
+            end
+        end
+
+        local is_in_view = is_y
+                           and (scrolled_pos + content_h > 0)
+                           or (scrolled_pos + content_w > 0)
+
+        if is_in_view then
+            -- Add the spacing widget, but not before the first widget
+            if i > 1 and spacing_widget then
+                table.insert(result, base.place_widget_at(
+                    spacing_widget,
+                    -- The way how spacing is added for regular widgets
+                    -- and the `spacing_widget` is disconnected:
+                    -- The offset for regular widgets is added to `pos` one
+                    -- iteration _before_ the one where the widget is actually
+                    -- placed.
+                    -- Because of that, the placement for the spacing widget
+                    -- needs to substract that offset to be placed right after
+                    -- the previous regular widget.
+                    math.floor(is_y and content_x or (content_x - spacing)),
+                    math.floor(is_y and (content_y - spacing) or content_y),
+                    math.floor(is_y and content_w or spacing),
+                    math.floor(is_y and spacing or content_h)
+                ))
+            end
+
+            table.insert(result, base.place_widget_at(
+                w,
+                math.floor(content_x),
+                math.floor(content_y),
+                math.floor(content_w),
+                math.floor(content_h)
+            ))
+        end
+    end
+
+    return result
+end
+
+function overflow:before_draw_children(_, cr, width, height)
+    -- Clip drawing for children to the space we're allowed to draw in
+    cr:rectangle(0, 0, width, height)
+    cr:clip()
```

Comment:
This clips what is drawn, but the input code does not know that you clipped something. Assuming you place a widget partially outside of the container (I have not yet entirely understood `:layout()` yet and am not 100% sure what is done here), it will still be possible to click into the not-visible parts of the widget.

(The scroll container does dark magic that results in *no* mouse clicks being possible; not really much better.)

Reply:
Your assumption is correct. Another user reported this as well, as a general problem with clipped drawing. If I remember them correctly, clicks fire for any non-visible part of a widget inside a container with some `cr:clip()`, even if that part doesn't extend outside the container's geometry.

For the specific case here, this patch supposedly helps: https://github.com/xinhaoyuan/awesome/commit/384b9eaa597396f53e944865c61ca0edd5e4c509, but I guess there are other use cases that need things to work like they do currently.

Reply:
If clipping of child widgets is something that is needed... well, I guess it might (in theory) be possible to extend `wibox.widget.base.place_widget_*` with a `shape` parameter.

However, the result can easily end up being ugly: Doing hit tests against a `gears.shape` requires creating a temporary cairo surface and context, applying the shape, and then checking with something like `cr:in_fill(x, y)`. This sounds like a "nice" way to end up with high memory usage and/or slow code.

Reply:
I don't think we would need a separate clip shape per child widget. A single method `widget.clip_shape` on the layout/parent should be enough.

And the special case of `widget.clip_shape == gears.rectangle` (or by checking an additional property) could then be optimized to skip the temporary cairo surface and use the patch linked above.